src/coordination/azure/Akka.Coordination.Azure/Internal/AzureApiImpl.cs

```diff
@@ -0,0 +1,338 @@
+using Newtonsoft.Json;
+
+namespace Akka.Coordination.Azure.Internal
+{
+    internal class AzureApiImpl: IAzureApi
+    {
+        private readonly AzureLeaseSettings _settings;
+        private readonly ILoggingAdapter _log;
+        private readonly Lazy<BlobContainerClient> _containerClient;
+
+        public AzureApiImpl(ActorSystem system, AzureLeaseSettings settings)
+        {
+            _settings = settings;
+            _log = Logging.GetLogger(system, GetType());
+
+            _containerClient = new Lazy<BlobContainerClient>(() =>
+            {
+                var client = settings.AzureCredential != null && settings.ServiceEndpoint != null
+                    ? new BlobContainerClient(settings.ServiceEndpoint, settings.AzureCredential,
+                        settings.BlobClientOptions)
+                    : new BlobContainerClient(settings.ConnectionString, settings.ContainerName);
+
+                var response = client.Exists();
+                if (!response.Value)
+                    client.Create();
+                
+                return client;
+            });
+        }
+        
+        [InternalApi]
+        public async Task<LeaseResource> ReadOrCreateLeaseResource(string name)
+        {
+            // TODO: backoff retry
+            var maxTries = 5;
+            var tries = 0;
+            while (true)
+            {
+                if (await LeaseResourceExists(name))
+                {
+                    var olr = await GetLeaseResource(name);
+                    if (olr != null)
+                    {
+                        _log.Debug("{0} already exists. Returning {1}", name, olr);
+                        return olr;
+                    }
+                }
+                else
+                {
+                    _log.Info("lease {0} does not exist, creating", name);
+                    var olr = await CreateLeaseResource(name);
+                    if (olr != null)
+                        return olr;
+                }
+                
+                tries++;
+                if (tries >= maxTries)
+                    throw new LeaseException($"Unable to create or read lease after {maxTries} tries");
+            }
+            
+        }
+
+        [InternalApi]
+        public async Task<Either<LeaseResource, LeaseResource>> UpdateLeaseResource(
+            string leaseName, string ownerName, ETag version, DateTimeOffset? time = null)
+        {
+            var cts = new CancellationTokenSource(_settings.BodyReadTimeout);
+            try
+            {
+                _log.Debug("Updating {0}", leaseName);
+                
+                var blobClient = _containerClient.Value.GetBlobClient(leaseName);
+                var leaseBody = new LeaseBody(ownerName, time);
+                var options = new BlobUploadOptions
+                {
+                    Conditions = new BlobRequestConditions
+                    {
+                        IfMatch = version
```

Comment:
So what's the importance of this? Make sure we can't override someone else's lease with our own?

Reply:
This is the concurrency lock, we can only update the lease if our ETag matches the server ETag, else the lease was taken by someone else